### PR TITLE
feat(apis_entities): add a default search field to the entity filtersets

### PR DESCRIPTION
**Pull request by b1rger**
This default search uses `generic.helpers.generate_search_filter` to
either search through all the Char and Textfields or through the fields
the model has defined in `_default_search_fields`.
It uses the same logic to list the searched fields in the help text of
the searchfield.

### CODE REVIEW

File: apis_core/apis_entities/filtersets.py [hunk 1]
```diff
@@ -22,6 +22,30 @@
...
+                fields = model._default_search_fields
+            else:
+                modelfields = model._meta.fields
+                fields = [
+                    field.name
+                    for field in modelfields
+                    if isinstance(field, (models.CharField, models.TextField))
+                ]
+            fields = ", ".join(fields)
+            self.extra["help_text"] = f"Search in the fields: {fields}"
```
Comment (koeaw) [inline apis_core/apis_entities/filtersets.py:43]: I'd change this to `f"Search in fields: {fields}"`

**Comment by koeaw** [comment]
This is really interesting!

Having tested it on Frischmuth, which already has a `search` field/filter for several entities, I'm wondering about a few things:
* Is there a way to turn this off? Or could there be? (on a per-model basis, or altogether) Or could this be opt-in rather than opt-out?
* Could this be modified so a custom field label and `help_text` could be provided? (in our case, we'd want to provide our own German strings)
* The list of field names would ideally be made up of verbose names (to match the labelling in the rest of the form/elsewhere)
* This (unexpectedly) overrides Frischmuth's custom filters because we use the same field/filter name, `search`. Could the hierarchy of filters be changed so project-specific filtes trump this generic search filter? And/or could the name be changed to something less likely to clash with custom filters (i.e. something less generic)?

Scenarios for the last point: A project wants to make use of this default search filter across models, but would need to override it in a few cases with a custom definition. Or: a project has its own primary `search` filter per model, but would _additionally_ like to make the generic search filter available.

E.g. in Frischmuth, we want our main "Search" field to be limited to very specific fields + use fuzzy search. But, knowing this generic filter exists, I think I'd like to make it available as well, as secondary, less "precise" filter, in case that's useful.

**Comment by b1rger** [comment]
>   Is there a way to turn this off? Or could there be? (on a per-model basis, or altogether) Or could this be opt-in rather than opt-out?

You can turn this off by providing your own filterset - I think a search field should be part of the default filterset.

>  Could this be modified so a custom field label and `help_text` could be provided? (in our case, we'd want to provide our own German strings)

Yes, you can simply inherit from `ModelSearchFilter` and override the `help_text` in the init

>  The list of field names would ideally be made up of verbose names (to match the labelling in the rest of the form/elsewhere)

ack

>   This (unexpectedly) overrides Frischmuth's custom filters because we use the same field/filter name, `search`. Could the hierarchy of filters be changed so project-specific filtes trump this generic search filter? And/or could the name be changed to something less likely to clash with custom filters (i.e. something less generic)?

This is simply how object inheritance works in Python. You can override the `__init__` to make your custom search be the one thats picked. We can not put the default search into the filterset as a class variable, because there we don't have access to the `Meta.model`.


> Scenarios for the last point: A project wants to make use of this default search filter across models, but would need to override it in a few cases with a custom definition. Or: a project has its own primary `search` filter per model, but would _additionally_ like to make the generic search filter available.

Thats possible using object inheritance


> E.g. in Frischmuth, we want our main "Search" field to be limited to very specific fields 

Thats exactly what this search does

> + use fuzzy search.

As already discussed, this can be implemented in the `generic_search_filter` helper

> But, knowing this generic filter exists, I think I'd like to make it available as well, as secondary, less "precise" filter, in case that's useful.

**Comment by b1rger** [comment]
> > E.g. in Frischmuth, we want our main "Search" field to be limited to very specific fields
> 
> Thats exactly what this search does

(... once #946 gets reviewed and merged)

**Comment by koeaw** [comment]
When I last tested this (end of July), I eventually managed to work around this change clashing with/overwriting the custom `search` fields already in place in Frischmuth.

I.e. this worked for me – with limitations.

What I'm missing:
* `verbose_names` instead of actual field names for the list of fields in the search field's label
* `ArrayFields` which have `CharFields` or `TextFields` as `base_field` being covered as well
*  (?*) `ManyToManyFields` which reference `CharFields` or `TextFields` being covered as well

(*I'm not _entirely_ sure anymore if I tested this in the end, but I feel like I did...)

**Review by koeaw — APPROVED** [review]

**Pull request merged**